platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c

```diff
@@ -0,0 +1,341 @@
+    1, /* supports UART (Asynchronous) mode */
+    0, /* supports Synchronous Master mode */
+    0, /* supports Synchronous Slave mode */
+    0, /* supports UART Single-wire mode */
+    0, /* supports UART IrDA mode */
+    0, /* supports UART Smart Card mode */
+    0, /* Smart Card Clock generator available */
+    0, /* RTS Flow Control available */
+    0, /* CTS Flow Control available */
+    0, /* Transmit completed event: \ref ARM_USART_EVENT_TX_COMPLETE */
+    0, /* Signal receive character timeout event: \ref ARM_USART_EVENT_RX_TIMEOUT */
+    0, /* RTS Line: 0=not available, 1=available */
+    0, /* CTS Line: 0=not available, 1=available */
+    0, /* DTR Line: 0=not available, 1=available */
+    0, /* DSR Line: 0=not available, 1=available */
+    0, /* DCD Line: 0=not available, 1=available */
+    0, /* RI Line: 0=not available, 1=available */
+    0, /* Signal CTS change event: \ref ARM_USART_EVENT_CTS */
+    0, /* Signal DSR change event: \ref ARM_USART_EVENT_DSR */
+    0, /* Signal DCD change event: \ref ARM_USART_EVENT_DCD */
+    0  /* Signal RI change event: \ref ARM_USART_EVENT_RI */
+};
+
+typedef struct {
+    nrfx_uarte_t        uarte;
+    nrfx_uarte_config_t config;
+    size_t              tx_count;
+    size_t              rx_count;
+} UARTx_Resources;
+
+//
+//   Functions
+//
+
+ARM_DRIVER_VERSION ARM_USART_GetVersion(void)
+{
+    return DriverVersion;
+}
+
+ARM_USART_CAPABILITIES ARM_USART_GetCapabilities(void)
+{
+    return DriverCapabilities;
+}
+
+int32_t ARM_USARTx_Initialize(ARM_USART_SignalEvent_t cb_event,
+                              UARTx_Resources *uart_resources)
+{
+    ARG_UNUSED(cb_event);
+
+    nrfx_err_t err_code = nrfx_uarte_init(&uart_resources->uarte,
+                                          &uart_resources->config,
+                                          NULL);
+    return (err_code == NRFX_SUCCESS) ? ARM_DRIVER_OK
+                                      : ARM_DRIVER_ERROR_BUSY;
+}
+
+int32_t ARM_USARTx_Uninitialize(UARTx_Resources *uart_resources)
+{
+    nrfx_uarte_uninit(&uart_resources->uarte);
+    return ARM_DRIVER_OK;
+}
+
+int32_t ARM_USARTx_PowerControl(ARM_POWER_STATE state,
+                                UARTx_Resources *uart_resources)
+{
+    ARG_UNUSED(uart_resources);
+
+    switch (state) {
+    case ARM_POWER_OFF:
+    case ARM_POWER_LOW:
+        return ARM_DRIVER_ERROR_UNSUPPORTED;
+    case ARM_POWER_FULL:
+        /* Nothing to be done */
+        return ARM_DRIVER_OK;
+    /* default:  The default is not defined intentionally to force the
+     *           compiler to check that all the enumeration values are
+     *           covered in the switch.*/
+    }
+}
+
+int32_t ARM_USARTx_Send(const void *data, uint32_t num,
+                        UARTx_Resources *uart_resources)
+{
+    nrfx_err_t err_code = nrfx_uarte_tx(&uart_resources->uarte, data, num);
+    if (err_code != NRFX_SUCCESS) {
+        return ARM_DRIVER_ERROR_BUSY;
```

Comment:
This might hide more serious errors, like NRFX_ERROR_INTERNAL.

Reply:
Right. Will change this.

Reply:
Should ARM_USARTx_Initialize() get the same treatment?

Reply:
No, the only error in `ARM_USARTx_Initialize` can be that there is an attempt to initialize an already initialized nrfx_uarte driver instance.
In `ARM_USARTx_Send` and `ARM_USARTx_Receive`, the driver can be busy with a previous transfer (`ARM_DRIVER_ERROR_BUSY` is returned then) or some other error can occur, like invalid buffer passed (not placed in RAM) or an internal nrfx_uarte driver error, then `ARM_DRIVER_ERROR` is returned.